File: vat_compliance/vat_compliance/report/mushak_6_3/mushak_6_3.js
// Copyright (c) 2025, na and contributors
// For license information, please see license.txt

frappe.query_reports["Mushak 6_3"] = {
	filters: [
		{
			fieldname: "from_date",
			label: "From Date",
			fieldtype: "Date",
			default: frappe.datetime.month_start(),
		},
		{
			fieldname: "to_date",
			label: "To Date",
			fieldtype: "Date",
			default: frappe.datetime.month_end(),
		},
		{
			fieldname: "company",
			label: "Company",
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			read_only: 1,
		},
		{
			fieldname: "customer",
			label: "Customer",
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "sales_invoice",
			label: "Sales Invoice",
			fieldtype: "Link",
			options: "Sales Invoice",
		},
		{
			fieldname: "status",
			label: "Status",
			fieldtype: "Select",
			options: [
				"",
				"Draft",
				"Return",
				"Credit Note Issued",
				"Submitted",
				"Paid",
				"Partly Paid",
				"Unpaid",
				"Unpaid and Discounted",
				"Partly Paid and Discounted",
				"Overdue and Discounted",
				"Overdue",
				"Cancelled",
				"Internal Transfer",
			].join("\n"),
		},
	],
	get_datatable_options(options) {
		return Object.assign(options, {
			cellHeight: 36,
		});
	},
	onload(report) {
		// Add Download All button
		this.setup_download_all_button(report);
	},

	setup_download_all_button(report) {
		// Add Download All button to the page
		report.page.add_inner_button(__("Download All"), () => {
			this.download_all_invoices(report);
		});
	},

	download_all_invoices(report) {
		// Show loading indicator
		const filters = report.get_values();

		window.open(
			frappe.urllib.get_full_url(
				"/api/method/vat_compliance.vat_compliance.report.mushak_6_3.mushak_6_3.download_all_invoices?filters=" +
					encodeURIComponent(JSON.stringify(filters))
			)
		);
	},
};

File: vat_compliance/vat_compliance/report/mushak_6_3/mushak_6_3.py
# Copyright (c) 2025, na and contributors
# For license information, please see license.txt

import json

import frappe
from frappe import _


def execute(filters=None):
	if not filters:
		filters = {}

	columns = get_columns()
	data = get_data(filters)

	return columns, data


def get_columns():
	return [
		{"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 110},
		{
			"label": _("Invoice No"),
			"fieldname": "name",
			"fieldtype": "Link",
			"options": "Sales Invoice",
			"width": 200,
		},
		{
			"label": _("Customer"),
			"fieldname": "customer",
			"fieldtype": "Link",
			"options": "Customer",
			"width": 180,
		},
		{
			"label": _("Company"),
			"fieldname": "company",
			"fieldtype": "Link",
			"options": "Company",
			"width": 150,
		},
		{"label": _("Total (Excl. VAT)"), "fieldname": "net_total", "fieldtype": "Currency", "width": 130},
		{
			"label": _("Total VAT"),
			"fieldname": "total_taxes_and_charges",
			"fieldtype": "Currency",
			"width": 120,
		},
		{"label": _("Grand Total"), "fieldname": "grand_total", "fieldtype": "Currency", "width": 130},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 100},
		{"label": _("Download 6.3"), "fieldname": "download", "fieldtype": "Data", "width": 130},
	]


def get_data(filters):
	conditions = []
	values = {}

	if filters.get("from_date"):
		conditions.append("posting_date >= %(from_date)s")
		values["from_date"] = filters.get("from_date")

	if filters.get("to_date"):
		conditions.append("posting_date <= %(to_date)s")
		values["to_date"] = filters.get("to_date")

	if filters.get("company"):
		conditions.append("company = %(company)s")
		values["company"] = filters.get("company")

	if filters.get("customer"):
		conditions.append("customer = %(customer)s")
		values["customer"] = filters.get("customer")

	if filters.get("sales_invoice"):
		conditions.append("name = %(sales_invoice)s")
		values["sales_invoice"] = filters.get("sales_invoice")

	if filters.get("status"):
		conditions.append("status = %(status)s")
		values["status"] = filters.get("status")
	else:
		conditions.append("docstatus = 1")

	where_clause = " AND ".join(conditions)

	query = f"""
        SELECT
            name,
            posting_date,
            customer,
            company,
            net_total,
            total_taxes_and_charges,
            grand_total,
            status
        FROM `tabSales Invoice`
        WHERE {where_clause}
        ORDER BY posting_date DESC
    """

	invoices = frappe.db.sql(query, values, as_dict=True)

	for inv in invoices:
		inv["download"] = (
			f"<a class='btn btn-xs btn-success' "
			f"href='/printview?doctype=Sales%20Invoice&name={inv.name}"
			f"&trigger_print=1' target='_blank'>"
			f"<i class='fa fa-download'></i></a>"
		)

	return invoices


@frappe.whitelist()
def download_all_invoices(filters):
	"""
	Generate PDFs for all invoices and create a zip file for download
	"""
	import zipfile
	from io import BytesIO

	if isinstance(filters, str):
		filters = frappe.parse_json(filters)

	invoices = get_data(filters)
	if not invoices:
		frappe.throw(_("No invoices found for the selected filters"))

	try:
		zip_buffer = BytesIO()

		with zipfile.ZipFile(zip_buffer, "w", zipfile.ZIP_DEFLATED) as zipf:
			for invoice in invoices:
				try:
					pdf_content = frappe.get_print(
						doctype="Sales Invoice",
						name=invoice.get("name"),
						as_pdf=True,
						pdf_generator="wkhtmltopdf",
					)
					filename = f"Mushak_6_3_{invoice['name']}.pdf"
					zipf.writestr(filename, pdf_content)

				except Exception as e:
					frappe.log_error(
						f"Error generating PDF for invoice {invoice['name']}: {e!s}",
						"Download All Invoices",
					)
					continue

		zip_buffer.seek(0)

		file_name = f"mushak_6_3_invoices_{frappe.utils.now_datetime().strftime('%Y%m%d_%H%M%S')}.zip"

		frappe.response["filename"] = file_name
		frappe.response["filecontent"] = zip_buffer.getvalue()
		frappe.response["type"] = "download"
		frappe.response["content_type"] = "application/zip"

	except Exception as e:
		frappe.log_error(f"Error in download_all_invoices: {e!s}", "Download All Invoices")
		frappe.throw(_("Error generating zip file: {0}").format(str(e)))
